yarn-project/prover-client/src/proving_broker/broker_prover_facade.ts

```diff
@@ -53,77 +75,315 @@ export class BrokerCircuitProverFacade implements ServerCircuitProver {
-    await this.broker.enqueueProvingJob({
+    const jobStatus = await this.broker.enqueueProvingJob({
       id,
       type,
       inputsUri,
       epochNumber,
     });
 
-    this.log.verbose(
-      `Sent proving job to broker id=${id} type=${ProvingRequestType[type]} epochNumber=${epochNumber}`,
-      {
-        provingJobId: id,
-        provingJobType: ProvingRequestType[type],
-        epochNumber,
-        inputsUri: truncate(inputsUri),
-      },
-    );
-
-    // notify broker of cancelled job
+    // Create a promise for this job id, regardless of whether it was enqueued at the broker
+    // The running promise will monitor for the job to be completed and resolve it either way
+    const promise = promiseWithResolvers<ProvingJobResultsMap[T]>();
     const abortFn = async () => {
       signal?.removeEventListener('abort', abortFn);
       await this.broker.cancelProvingJob(id);
     };
+    const job: ProvingJob = {
+      id,
+      type,
+      promise,
+      abortFn,
+      signal,
+    };
+    this.jobs.set(id, job);
+
+    // If we are here then the job was successfully accepted by the broker
+    // the returned status is for before any action was performed
+    if (jobStatus.status === 'not-found') {
+      // Job added for the first time
+      // notify the broker if job is aborted
+      signal?.addEventListener('abort', abortFn);
+
+      this.log.verbose(
+        `Job enqueued with broker id=${id} type=${ProvingRequestType[type]} epochNumber=${epochNumber}`,
+        {
+          provingJobId: id,
+          provingJobType: ProvingRequestType[type],
+          epochNumber,
+          inputsUri: truncate(inputsUri),
+          numOutstandingJobs: this.jobs.size,
+        },
+      );
+    } else if (jobStatus.status === 'fulfilled' || jobStatus.status === 'rejected') {
+      // Job was already completed by the broker
+      // No need to notify the broker on aborted job
+      job.abortFn = undefined;
+      this.log.verbose(
+        `Job already completed when sent to broker id=${id} type=${ProvingRequestType[type]} epochNumber=${epochNumber}`,
+        {
+          provingJobId: id,
+          provingJobType: ProvingRequestType[type],
+          epochNumber,
+          inputsUri: truncate(inputsUri),
+        },
+      );
 
-    signal?.addEventListener('abort', abortFn);
-
-    try {
-      // loop here until the job settles
-      // NOTE: this could also terminate because the job was cancelled through event listener above
-      const result = await retryUntil(
-        async () => {
-          try {
-            return await this.broker.waitForJobToSettle(id);
-          } catch (err) {
-            // waitForJobToSettle can only fail for network errors
-            // keep retrying until we time out
-          }
+      // Job was not enqueued. It must be completed already, add to our set of already completed jobs
+      this.jobsToRetrieve.add(id);
```

Comment:
On the new else branch (now that we have status) we technically have the result in the status object and can resolve immediately, but an optimization for another time :)